ui/components/TracePage/DetailRenderer.tsx

```diff
@@ -0,0 +1,219 @@
+import { useToast } from "@chakra-ui/react";
+import { getFormattedName } from "/components/TracePage/CallName";
+import {
+  FlattenedFStringPart,
+  flattenFString,
+  FString,
+  RawFStringPart,
+} from "/components/TracePage/FString";
+import classNames from "classnames";
+import { Component, useMemo } from "react";
+import { CaretDown } from "phosphor-react";
+
+type JsonChild =
+  | { type: "array"; values: unknown[] }
+  | { type: "object"; values: [string, unknown][] }
+  | { type: "value"; value: unknown; fstring?: FlattenedFStringPart[] };
+
+const getStructuralType = (data: unknown) => {
+  if (typeof data === "object" && data && !Array.isArray(data)) return "object";
+  if (Array.isArray(data)) return "array";
+  return "value";
+};
+
+const TypeIdentifiers = {
+  object: <span className="shrink-0 font-mono mr-[8px]">{"{}"}</span>,
+  array: <span className="shrink-0 font-mono mr-[8px]">{"[]"}</span>,
+  value: null,
+};
+
+interface ArrayRendererProps {
+  values: unknown[];
+}
+
+interface ArrayRendererState {
+  expanded: boolean[];
+}
+
+class ArrayRenderer extends Component<ArrayRendererProps, ArrayRendererState> {
+  constructor(props: ArrayRendererProps) {
+    super(props);
+    this.state = {
```

Comment:
Please switch to a function-based component and use the `setState` hook to manage `expanded`.

Reply:
Same for the other class-based components.

Reply:
Thanks! https://github.com/oughtinc/ice/commit/a70a48a0479663e06303675ded799b64e07b5850 and f0a4b75703585942aafda7efdc1efd27641d324d address this.